_posts/2020-07-29-springboot-logging.md

```diff
@@ -0,0 +1,336 @@
+
+
+| Property        | What It Means           | Value If Not Set  |
+| --------------------- |:-------------| :-------------:|
+| logging.file.max-size      | total size of log archive | 10 Mb |
+| logging.file.max-history      | how many days' rotated log files to be kept      |   7 Days |
+| logging.file.total-size-cap | total size of log archives. Backups are deleted when the total size of log archives exceeds that threshold.      |   Not Specified |
+| logging.file.clean-history-on-start | force log archive cleanup on application startup      |    false |
+
+
+We can apply the same customization in a separate file which we will see in the next section. 
+
+### Isolate Logging Configuration From Application With logback Configuration
+ 
+We can isolate the log configuration from the application by specifying the configuration in logback.xml or logback-spring in XML or groovy syntax. Spring recommends using the names with spring-like logback-spring.xml or logback-spring.groovy.
+The configuration is comprised of appender element inside a root configuration tag. The pattern is specified inside an encoder element :
+
+```xml
+<configuration >
+  <include
+    resource="/org/springframework/boot/logging/logback/base.xml" />
+  <appender name="STDOUT"
+    class="ch.qos.logback.core.ConsoleAppender">
+    <encoder>
+      <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n
+      </pattern>
+    </encoder>
+  </appender>
+```
+
+Logback uses a configuration library - Joran so we will see these logs during application startup if we set a debug property in the configuration tag to true.
+
+```xml
+<configuration debug="true">
+```
+
+## Making Our Logging Useful
+We need to capture relevant information in our logs for our logging to be useful. Let us look at few of those:
+### What do we log
+The utility of logs depends on the information we capture during logging. Some common usages include data written by logs are :  
+1. Request/Response payload: Log the request and response payload of an API. This forms the starting point of any abnormal conditions in the system behavior. For instance, we check whether the input is valid, if the status of other sub-systems were healthy, etc.
+2. Headers: We log important request and response headers, for example, the JWT authorization header, headers giving context information like model of the device that sent the request, geography of the user.
+3. Method entry/exit: Similar to API, we might log the inputs and output of important methods to ensure that all the statements within the method executed successfully.
+4. Useful checkpoints within our source code to indicate the occurrence of any important event, for example, payment successful with payment reference, the message posted to a queue, etc.
+5. Exceptions are often captured in error logs with a complete stack trace and are a valuable component for any system diagnosis. 
+
+### Add Contextual Information
```

Comment:
Link this to `logback-spring.xml` and move this section into a subsection of "Customizing the Logger in Spring Boot`